src/node_crypto.cc

```diff
@@ -2610,7 +2614,7 @@ void CipherBase::CommonInit(const char* cipher_type,
                             "Failed to initialize cipher");
   }
 
-  if (IsSupportedAuthenticatedMode(mode)) {
+  if (IsSupportedAuthenticatedMode(mode) || IsAEADChacha20Poly1305(cipher)) {
```

Comment:
Shouldn't this check be in `IsSupportedAuthenticatedMode()`? The purpose of that function seems to be to remove the list of specific checks from places where AEAD modes need to be treated specially. I realize AEAD-chacha20-poly1305 is a bit odd in that it was not specified as a mode that can combine an arbitrary symetric cipher with an arbitrary MAC, it looks like it was speced for just a single cipher and MAC in https://tools.ietf.org/html/rfc7539#section-2.8. What `mode` is returned by `EVP_CIPHER_mode(cipher)` for aead-chacha20-poly1305? Is the mode insufficient to see that the cipher is running in AEAD mode? If it is, then would it be cleaner to modify `IsSupportedAuthenticatedMode()` to take either a cipher, or a cipher and a mode, as an argument so that all the checks for AEAD don't need a specific `IsAEADChacha20Poly1305()` check added beside them?

Reply:
Thank you very much about your review! The value EVP_CIPHER_mode(cipher) for aead-chacha20-poly1305 is 0 which is not a 'valid' aead mode, so I refactored the `IsSupportedAuthenticatedMode` function and making it to take a cipher as param.